src/transformers/models/mt5/modeling_mt5.py

```diff
@@ -14,29 +14,1254 @@
 # limitations under the License.
 """ PyTorch mT5 model."""
 
-from ...utils import logging
-from ..t5.modeling_t5 import T5EncoderModel, T5ForConditionalGeneration, T5Model
+import copy
+import math
+import os
+import warnings
+from typing import Optional, Tuple, Union
+
+import torch
+from torch import nn
+from torch.nn import CrossEntropyLoss
+from torch.utils.checkpoint import checkpoint
+
+from ...activations import ACT2FN
+from ...modeling_outputs import (
+    BaseModelOutput,
+    BaseModelOutputWithPastAndCrossAttentions,
+    Seq2SeqLMOutput,
+    Seq2SeqModelOutput,
+)
+from ...modeling_utils import PreTrainedModel
+from ...pytorch_utils import find_pruneable_heads_and_indices, prune_linear_layer
+from ...utils import (
+    DUMMY_INPUTS,
+    DUMMY_MASK,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    is_torch_fx_proxy,
+    logging,
+    replace_return_docstrings,
+)
+from ...utils.model_parallel_utils import assert_device_map, get_device_map
 from .configuration_mt5 import MT5Config
 
 
 logger = logging.get_logger(__name__)
 
-_CONFIG_FOR_DOC = "T5Config"
-_TOKENIZER_FOR_DOC = "T5Tokenizer"
+_CONFIG_FOR_DOC = "MT5Config"
+_TOKENIZER_FOR_DOC = "MT5Tokenizer"
+_CHECKPOINT_FOR_DOC = "mt5-small"
 
 
-class MT5Model(T5Model):
-    r"""
-    This class overrides [`T5Model`]. Please check the superclass for the appropriate documentation alongside usage
-    examples.
+PARALLELIZE_DOCSTRING = r"""
+    This is an experimental feature and is a subject to change at a moment's notice.
+
+    Uses a device map to distribute attention modules of the model across several devices. If no device map is given,
+    it will evenly distribute blocks across all devices.
+
+    Args:
+        device_map (`Dict[int, list]`, optional, defaults to None):
+            A dictionary that maps attention modules to devices. Note that the embedding module and LMHead are always
+            automatically mapped to the first device (for esoteric reasons). That means that the first device should
+            have fewer attention modules mapped to it than other devices. For reference, the mt5 models have the
+            following number of attention modules:
+
+                - mt5-small: 6
+                - mt5-base: 12
+                - mt5-large: 24
+                - mt5-3b: 24
```

Comment:
Hey, thanks for making this PR! I think mt3 has different model names than T5, it is instead mt5-xl, mt5-xxl rather than 3b, 11b [found here](https://huggingface.co/docs/transformers/model_doc/mt5). 

Moreover, I think their number of attention modules might also be different, seen [here](https://github.com/google-research/text-to-text-transfer-transformer/blob/main/released_checkpoints.md#t511) since mt5 is based off of t5.1.1 rather than t5 as the authors of mt5 note [here](https://arxiv.org/pdf/2010.11934.pdf).

Reply:
Yeah, thanks for mentioning it, I have updated those model names but still looking for the number of attention modules for mt5. @sgugger, can you please help me with this? Thanks,